### PR TITLE
resharding: auto-reverse replication

### PR DESCRIPTION
**Pull request by sougou**
MigrateServedTypes for master automatically does the prep work for
reversing replication. If the flag is specified, it reverses the
direction thereby allowing you to migrate back to the original
masters.

If the flag was not originally specified, then it only does the prep
work, without actually starting the replication. You can later
reissue the Migrate command, which will then start the vreplication.

Signed-off-by: Sugu Sougoumarane <ssougou@gmail.com>

### CODE REVIEW

**Comment by tirsen** [comment]
WHOA!

File: go/vt/vtctl/vtctl.go [hunk 1]
```diff
@@ -1684,6 +1684,7 @@ func commandMigrateServedTypes(ctx context.Context, wr *wrangler.Wrangler, subFl
 	reverse := subFlags.Bool("reverse", false, "Moves the served tablet type backward instead of forward. Use in case of trouble")
 	skipReFreshState := subFlags.Bool("skip-refresh-state", false, "Skips refreshing the state of the source tablets after the migration, meaning that the refresh will need to be done manually, replica and rdonly only)")
 	filteredReplicationWaitTime := subFlags.Duration("filtered_replication_wait_time", 30*time.Second, "Specifies the maximum time to wait, in seconds, for filtered replication to catch up on master migrations")
+	reverseReplication := subFlags.Bool("reverse_replication", false, "For master migration, enabling this flag reverses replication allowing which allows you to rollback")
```
Comment (tirsen) [inline go/vt/vtctl/vtctl.go]: Remove "allowing"

File: go/vt/wrangler/keyspace.go [hunk 2]
```diff
@@ -388,12 +410,18 @@ func (wr *Wrangler) masterMigrateServedType(ctx context.Context, keyspace string
...
 	// We've reached the point of no return. Freeze the tablet control records in the source masters.
-	if err := wr.freezeMasterMigrateServedType(ctx, sourceShards); err != nil {
+	if err := wr.updateFrozenFlag(ctx, sourceShards, true); err != nil {
 		wr.cancelMasterMigrateServedTypes(ctx, sourceShards)
 		return err
 	}
 
 	// Phase 2
+	// Always setup reverse replication. We'll start it later if reverseReplication was specified.
+	// This will allow someone to reverse the replication later if they change their mind.
```
Comment (tirsen) [inline go/vt/wrangler/keyspace.go]: Ah so even if you don't pass the `reverse_replication` flag you can still start vreplication and revert the migration?

Comment (sougou) [reply go/vt/wrangler/keyspace.go]: Yes :)

Comment (sougou) [reply go/vt/wrangler/keyspace.go]: In fact, you should't start vreplication yourself. You should just reissue `MigrateServedTypes` with `-reverse_replication`.

**Comment by tirsen** [comment]
I think the safest option should be the default but this is a somewhat unproven option so maybe as a first step make it not the default and then we make it the default once it's been well tested in the field.

**Comment by tirsen** [comment]
Seems crazy not to save the metadata though.

**Comment by sougou** [comment]
Clarified with @tirsen: auto-reverse should not happen unless requested, but metadata should be created irrespective of the flag. This is the current implementation.

**Comment by demmer** [comment]
My preference would be somewhere in between...

Could we unconditionally save the master position metadata that you would need to start reverse replication, but don't set up the replication itself unless requested? It could be a new workflow type of "SavedReverseReplicationPos" or some such which would never actually be used to start the replication, but from which we could get the positions if needed.

**Comment by tirsen** [comment]
We don't currently use workflows so we wouldn't benefit from saving stuff there.

In my opinion safest option should always be the default. Based on that philosophy by default we should even start the reverse vreplication. That said, this is still a somewhat unproven feature so might need to see some more field testing first. As a first step setting up the vreplication but not starting it seems like a safe low risk option.

**Comment by rafael** [comment]
@demmer  idea makes sense. My main point with the discussion with @sougou is that current API is flexible but confusing from an user perspective.  I don't think is worth it to get stuck on that point though. 

I'm supportive of moving on and do something like @demmer  is suggesting in a future PR.

**Comment by tirsen** [comment]
For posterity, which approach did we end up merging? (Too lazy to read the code.)

**Comment by tirsen** [comment]
For posterity, which approach did we end up merging? (Too lazy to read the code.)

**Comment by tirsen** [comment]
For posterity, which approach did we end up merging? (Too lazy to read the code.)

**Comment by tirsen** [comment]
For posterity, which approach did we end up merging? (Too lazy to read the code.)

**Comment by sougou** [comment]
Always save metadata. Initiate auto-reverse only if requested.

**Pull request merged**